src/wikirec/model.py

```diff
@@ -327,3 +351,134 @@ def recommend(
+            
+    Returns
+    -------
+        book_weights : np.array
+            The normalized embedding vectors for each of the articles. 
+            
+            Shape of book_weights is (len(books), embedding_size)
+    
+    """
+    if os.path.isfile(path_to_json):
+        with open(path_to_json, "r") as fin:
+            books = [json.loads(l) for l in fin]
+    else:
+        raise Exception("Need to parse json for books.")
+        
+    # Find set of wikilinks for each book and convert to a flattened list
+    unique_wikilinks = list(chain(*[list(set(book[2])) for book in books]))
+    wikilinks = [link.lower() for link in unique_wikilinks]
+    to_remove = ['hardcover', 'paperback', 'hardback', 'e-book', 'wikipedia:wikiproject books', 'wikipedia:wikiproject novels']
+    wikilinks = [item for item in wikilinks if item not in to_remove]
+
+    # Limit to wikilinks that occur more than 4 times
+    wikilinks_counts = Counter(wikilinks)
+    wikilinks_counts = sorted(wikilinks_counts.items(), key = lambda x: x[1], reverse = True)
+    wikilinks_counts = OrderedDict(wikilinks_counts)
+    links = [t[0] for t in wikilinks_counts.items() if t[1] >= 4]
+    
+    # map books to their indices, and map links to indices as well 
+    book_index = {book[0]: idx for idx, book in enumerate(books)}
+    link_index = {link: idx for idx, link in enumerate(links)} 
+    
+    #Create data from pairs of (book, wikilink) for training the neural network embedding
+    pairs = []
+    for book in books:
+        title = book[0]
+        book_links = book[2]
+        # Iterate through wikilinks in book article
+        for link in book_links:
+            # Add index of book and index of link to pairs
+            if link.lower() in links:
+                pairs.append((book_index[title], link_index[link.lower()]))
+    pairs_set = set(pairs)
+    
+    # Neural network architecture
+    # Both inputs are 1-dimensional
+    book_input = Input(name = 'book', shape = [1])
+    link_input = Input(name = 'link', shape = [1])
+    
+    # Embedding the book (shape will be (None, 1, 50))
+    book_embedding = Embedding(name = 'book_embedding',
+                               input_dim = len(book_index),
+                               output_dim = embedding_size)(book_input)
+    
+    # Embedding the link (shape will be (None, 1, 50))
+    link_embedding = Embedding(name = 'link_embedding',
+                               input_dim = len(link_index),
+                               output_dim = embedding_size)(link_input)
+    
+    # Merge the layers with a dot product along the second axis 
+    # (shape will be (None, 1, 1))
+    merged = Dot(name = 'dot_product', normalize = True, 
+                 axes = 2)([book_embedding, link_embedding])
+    
+    # Reshape to be a single number (shape will be (None, 1))
+    merged = Reshape(target_shape = [1])(merged)
+    
+    model = Model(inputs = [book, link], outputs = merged)
```

Comment:
@victle, Figured out that this here is supposed to be `book_input` and `link_input` based on comparing it all to the original blogpost and your changes :)

Am making good progress on this and will try to have it done by tonight/early tomorrow. Thanks again for all this!